plugins/modules/packaging/language/cargo.py

```diff
@@ -0,0 +1,200 @@
+    def path(self):
+        return self._path
+
+    @path.setter
+    def path(self, path):
+        if path is not None and not os.path.isdir(path):
+            self.module.fail_json(msg="Path %s is not a directory" % path)
+        self._path = path
+
+    def _exec(
+        self, args, run_in_check_mode=False, check_rc=True, add_package_name=True
+    ):
+        if not self.module.check_mode or (self.module.check_mode and run_in_check_mode):
+            cmd = self.executable + args
+            rc, out, err = self.module.run_command(cmd, check_rc=check_rc)
+            return out
+        return ""
+
+    def get_installed(self):
+        cmd = ["install", "--list"]
+        data = self._exec(cmd, True, False, False)
+
+        package_regex = re.compile(r"^(\w+) v(.+):$")
+        installed = {}
+        for line in data.splitlines():
+            package_info = package_regex.match(line)
+            if package_info:
+                installed[package_info[1]] = package_info[2]
+
+        return installed
+
+    def install(self):
+        cmd = ["install"]
+        if self.name:
+            cmd.extend(self.name)
+        if self.path:
+            cmd.append("--root")
+            cmd.append(self.path)
+        if self.version:
+            cmd.append("--version")
+            cmd.append(self.version)
+        return self._exec(cmd)
+
+    def is_outdated(self, name):
+        installed_version = self.get_installed().get(name)
+
+        cmd = ["search", name, "--limit", "1"]
+        data = self._exec(cmd, True, False, False)
+
+        match = re.search(r'"(.+)"', data)
+        if match:
+            latest_version = match[1]
+
+        return installed_version != latest_version
+
+    def uninstall(self):
+        cmd = ["uninstall"]
+        if self.name:
+            cmd.extend(self.name)
+        return self._exec(cmd)
+
+
+def main():
+    arg_spec = dict(
+        name=dict(required=True, type="list", elements="str"),
+        path=dict(default=None, type="path"),
+        state=dict(default="present", choices=["present", "absent", "latest"]),
+        version=dict(default=None, type="str"),
+    )
+    module = AnsibleModule(argument_spec=arg_spec, supports_check_mode=True)
+
+    name = module.params["name"]
+    path = module.params["path"]
+    state = module.params["state"]
+    version = module.params["version"]
+
+    if not name:
+        module.fail_json(msg="Package name must be specified")
+
+    # Set LANG env since we parse stdout
+    module.run_command_environ_update = dict(
+        LANG="C", LC_ALL="C", LC_MESSAGES="C", LC_CTYPE="C"
+    )
+
+    cargo = Cargo(module, name=name, path=path, state=state, version=version)
+    changed = False
+    installed_packages = cargo.get_installed()
+    if state == "present":
+        if any(
+            (n not in installed_packages or version != installed_packages[n])
+            for n in name
+        ):
+            changed = True
+            cargo.install()
```

Comment:
It probably makes sense to pass only the packages that haven't been installed yet.

BTW, is there a way to specify the version as part of the name when calling `cargo install` (like when doing `pip install ansible==3.0.0`)? If yes, users will probably expect that the module accepts such combinations as well.

Reply:
> BTW, is there a way to specify the version as part of the name when calling `cargo install` (like when doing `pip install ansible==3.0.0`)? If yes, users will probably expect that the module accepts such combinations as well.

Unfortunately, cargo does not allow that. It installs a specific version using `--version` flag, which can only be specified once and applies to all listed packages.

Reply:
> It probably makes sense to pass only the packages that haven't been installed yet.

I agree. The current behavior is actually buggy, because if any packages were not installed, the module would update everything to the latest version due to the way cargo works.